### DISCUSSION TITLE
New integration of core and code generators modules

### DISCUSSION THREAD

**Discussion by zosrothko** [javacc/javacc-8#2] [Announcements]
Hi 

The javacc-8 repository has been improved with standalone independant git repo. There are now

1. core: the core is the JavaCC parser itself that takes care only about parsing the grammar and produce an AST.
2. codegen/cpp: the code generator for c++
3. codegen/java: the code generator for Java
4. codegen/cshap: the code generator for c#
5. codegen/js: the code generator for JavaScript

Currently, the csharp and js code generator are commented out from the codegen/pom.xml as the generated code does not compile

To building javacc-8, follow these steps
```
git clone git@github.com:javacc/javacc-8.git
cd javacc-8
git clone git@github.com:javacc/javacc-8-core.git core
git clone git@github.com:javacc/javacc-8-java.git codegen/java
git clone git@github.com:javacc/javacc-8-cpp.git codegen/cpp
git clone git@github.com:javacc/javacc-8-csharp.git codegen/csharp
git clone git@github.com:javacc/javacc-8-js.git codegen/js

mvn clean install

```

**Comment by zosrothko** [comment]
Here the log of the maven build. 
[install.log](https://github.com/javacc/javacc-8/files/9066229/install.log)

**Comment by new-javacc** [comment]
A coworker sent me this link regarding migrating git history to unrelated folders. Maybe we can use it to address some of @vlsi concerns?

https://gist.github.com/trongthanh/2779392

**Reply by zosrothko** [reply]
Will have a look next week

**Reply by kaikalur** [reply]
Others can also try - @vlsi if you want to. I'm very sloppy with git :( so I don't want to touch it as it might destroy everything,.

**Reply by vlsi** [reply]
@kaikalur, the attitude in javacc is not right. I suggested merging repositories long ago, and all I got was literally "wasting so much of my time" https://github.com/javacc/javacc-8-core/issues/1#issuecomment-1180745456
You may say whatever you want, however, the poll results in  https://github.com/javacc/javacc/discussions/242 are obvious: nobody wants splitting the code across repositories.

Please, do not ask me doing something for javacc project until the atmosphere in the project improves.

**Comment by zosrothko** [comment]
Why is this separated repositories are such a blolcking for you. All Java applications are build around the Java ecosystem with dependencies. Maven & Gradle are tools for managing such dependencies. When one have a Java application that depends on log4j for exemple, the socalled java application does  not include the source code of log4j!! That's the same with javacc-8. Each code generator in its own repository is only depending on the core artifact parametrized by the version number, as in any java application with dependencies. 

You can just clone the cpp code generator if you want, or only the core if you want. What is the issue?

An id you want all modules under a single directory umbrella, do as explained [here](https://github.com/javacc/javacc-8/discussions/2)

**Reply by vlsi** [reply]
> log4j

Please check log4j 2.x source code: https://github.com/apache/logging-log4j2
They have only one repository with 38 (!) modules, including:
* log4j-api <-- this jar the applications should use
* log4j-core <-- this is the core implementation, and the applications should not use it directly
* log4j-mongodb3 <-- this makes sense only if you need to log into mongodb3
* and so on.

> Why is this separated repositories are such a blolcking for you

Short answer: "working with separate repositories would consume **much more** time from **me**, and it will bring **zero benefits**".
Long answer: "many people explained why a single repository is much better, and you still keep asking <<why it is such a blocker?>>, it does not sound right for me".
From my point of view, the question "single repo vs multiple repos" is a no-brainer in the case of javacc. If you are so tied to your past decisions, then I doubt you will listen to other suggestions and ideas.

The poll results suggest that almost everyone suggests using only one repository: https://github.com/javacc/javacc/discussions/242

I already explained why having separate repositories is bad: https://github.com/javacc/javacc/discussions/238#discussioncomment-3109619. Would you please answer my questions in https://github.com/javacc/javacc/discussions/238#discussioncomment-3111848 ?

Another person listed the very same reasons for having only one repository: https://github.com/javacc/javacc/discussions/238#discussioncomment-3109644

**Nobody really explained the benefits of multiple repositories**.
To my best knowledge, the only reasoning behind separate repositories was "Maven". I am sure, that is very weak reasoning since even log4j manages to have 38 modules in a single Git repository, and they use Maven.

**Comment by kaikalur** [comment]
> Nobody really explained the benefits of multiple repositories.

I did many many times. The core is a mess and I don't have time to review PRs that touch core. So I want to control what goes in there whereas the code generators are open and free for all.

A compromise maybe to have and enforce OWNER file for core which currently will be only me until we refactor/clean it up for easier management.
